File: frontend/apps/ios/Pantopus/Features/Homes/PropertyDetails/PropertyDetailsViewModel.swift
//
//  PropertyDetailsViewModel.swift
//  Pantopus
//
//  Backs `PropertyDetailsView` (A.4 / A13.5). No backend — resolves a
//  `PropertyDetailsContent` from an injectable loader (defaults to the
//  bundled sample data) and projects it into a `.clean` / `.mismatch`
//  state based on whether external sources disagree.
//

import Foundation
import Observation

@Observable
@MainActor
final class PropertyDetailsViewModel {
    /// Currently displayed state.
    private(set) var state: PropertyDetailsState = .loading

    private let homeId: String
    private let loader: @Sendable (String) throws -> PropertyDetailsContent

    init(
        homeId: String,
        loader: @escaping @Sendable (String) throws -> PropertyDetailsContent = {
            PropertyDetailsSampleData.content(for: $0)
        }
    ) {
        self.homeId = homeId
        self.loader = loader
    }

    /// Initial load; no-op once content is resolved.
    func load() async {
        guard case .loading = state else { return }
        apply()
    }

    /// Retry after an error.
    func refresh() async {
        apply()
    }

    private func apply() {
        do {
            let content = try loader(homeId)
            state = content.banner == nil ? .clean(content) : .mismatch(content)
        } catch {
            state = .error(message: "Couldn't load property details. Pull to retry.")
        }
    }
}

File: frontend/apps/android/app/src/main/java/app/pantopus/android/ui/screens/homes/property_details/PropertyDetailsViewModel.kt
@file:Suppress("PackageNaming")

package app.pantopus.android.ui.screens.homes.property_details

import androidx.lifecycle.SavedStateHandle
import androidx.lifecycle.ViewModel
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import javax.inject.Inject

const val PROPERTY_DETAILS_HOME_ID_KEY = "homeId"

/** ViewModel for the read-mostly Property Details screen. */
@HiltViewModel
class PropertyDetailsViewModel
    @Inject
    constructor(
        savedStateHandle: SavedStateHandle,
    ) : ViewModel() {
        private val homeId: String =
            requireNotNull(savedStateHandle[PROPERTY_DETAILS_HOME_ID_KEY]) {
                "PropertyDetailsViewModel requires a '$PROPERTY_DETAILS_HOME_ID_KEY' nav arg."
            }

        private var loader: (String) -> PropertyDetailsContent = PropertyDetailsSampleData::contentFor
        private val _state = MutableStateFlow<PropertyDetailsUiState>(PropertyDetailsUiState.Loading)

        /** Observed state. */
        val state: StateFlow<PropertyDetailsUiState> = _state.asStateFlow()

        internal constructor(
            savedStateHandle: SavedStateHandle,
            loader: (String) -> PropertyDetailsContent,
        ) : this(savedStateHandle) {
            this.loader = loader
        }

        /** Initial load; no-op after content resolves. */
        fun load() {
            if (_state.value !is PropertyDetailsUiState.Loading) return
            apply()
        }

        /** Retry after an error. */
        fun refresh() {
            apply()
        }

        private fun apply() {
            _state.value =
                try {
                    val content = loader(homeId)
                    if (content.banner == null) {
                        PropertyDetailsUiState.Clean(content)
                    } else {
                        PropertyDetailsUiState.Mismatch(content)
                    }
                } catch (_: Exception) {
                    PropertyDetailsUiState.Error("Couldn't load property details. Pull to retry.")
                }
        }
    }
